### PR TITLE
cleanup: remove redundant method LookupAccountDataByAddress

### PR DESCRIPTION
**Pull request by icorderi**
## Summary

`LookupAccountDataByAddress` was only used in one place and was repeating the same logic as in `LookupAccount`. 
The difference between the two was that `LookupAccountDataByAddress` returned raw bytes instead  of decoding the account data. 

- `LookupAccountDataByAddress` was removed from the interface and the call site was  changed to used `LookupAccount` instead. 

## Test Plan

Existing tests.

### CODE REVIEW

**Review by AlgoAxel — DISMISSED** [review]

**Review by gmalouf — APPROVED** [review]
LGTM - easy to follow thinking, clean.

File: ledger/acctdeltas.go [hunk 1]
```diff
@@ -476,29 +476,13 @@ func (a *compactAccountDeltas) accountsLoadOld(tx trackerdb.TransactionScope) (e
 	}()
 	for _, idx := range a.misses {
 		addr := a.deltas[idx].address
-		ref, acctDataBuf, err := arw.LookupAccountDataByAddress(addr)
-		switch err {
-		case nil:
-			if len(acctDataBuf) > 0 {
-				persistedAcctData := &trackerdb.PersistedAccountData{Addr: addr, Ref: ref}
```
Comment (algorandskiy) [inline ledger/acctdeltas.go:483]: there is still a difference. this code does not set `Round` but `LookupAccount` does set. I did not check if this is important or not, but I suggest to either check or clear it here

**Comment by jsgranados** [comment]
Igancio and Pavel to follow up

File: ledger/acctdeltas.go [hunk 2]
```diff
@@ -476,29 +476,13 @@ func (a *compactAccountDeltas) accountsLoadOld(tx trackerdb.TransactionScope) (e
...
-				err = protocol.Decode(acctDataBuf, &persistedAcctData.AccountData)
-				if err != nil {
-					return err
-				}
-				a.updateOld(idx, *persistedAcctData)
-			} else {
-				// to retain backward compatibility, we will treat this condition as if we don't have the account.
-				a.updateOld(idx, trackerdb.PersistedAccountData{Addr: addr, Ref: ref})
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/acctdeltas.go:491]: oh, the second edge case with nil data. I'm pretty sure we do not have it in DB but right, it is better to have it covered.
a test case would be the following: there are 4 combinations of `new/old` that are translated to skip, delete, insert, update (there are test examples making such deltas). For each combo, use non-empty round, rowid (ref), data in `old` as a base case to get `accountsNewRoundImpl` output (return + mock writer). Then make (mutate) round, rowid, date to empty/nil values, and ensure the `accountsNewRoundImpl` produces the same results as the base case.

File: ledger/acctdeltas.go (hunk 1 above)
Comment (icorderi) [reply ledger/acctdeltas.go:483]: The `Round` is the round of the db at the point the "old value" was read. 
The old code wasn't setting it because it wasn't querying it, since it didn't seem to care about what round the old value was from. `LookupAccount` reads the round, because it's part of the data structure being read, and it might not be used by this caller, but it is used by others. 

This can't possible break code, because if any code was attempting to get meaning from the `Round` field in this "old value" entries, they all had `Round(0)`, implying that all this old values were read at genesis, which is flat out wrong.

File: ledger/acctdeltas.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (icorderi) [reply ledger/acctdeltas.go:491]: No it wont have `Ref: nil` @cce: 
```go
if len(buf) > 0 && rowid.Valid {
    data.Ref = sqlRowRef{rowid.Int64}
    err = protocol.Decode(buf, &data.AccountData)
    return err
}
```

As for @algorandskiy empty data column, for whatever meaning that might have, the  code can be changed to: 
 ```go
data.Ref = sqlRowRef{rowid.Int64}    
if len(buf) > 0 && rowid.Valid {
    err = protocol.Decode(buf, &data.AccountData)
    return err
}
```
But that changes the behavior of `LookupAccount` which is extensively tested directly and indirectly already.

File: ledger/acctdeltas.go [hunk 3]
```diff
@@ -476,29 +476,13 @@ func (a *compactAccountDeltas) accountsLoadOld(tx trackerdb.TransactionScope) (e
...
-			}
-		case sql.ErrNoRows:
```
[the remark this answers is not in the document]
Comment (icorderi) [reply ledger/acctdeltas.go:493]: No it's not the case. 

This shouldn't be possible, as far as the query for an account is concerned, a record is always returned.
```go
// this should never happen; it indicates that we don't have a current round in the acctrounds table.
if err == sql.ErrNoRows {
	// Return the zero value of data
	return fmt.Errorf("unable to query account data for address %v : %w", addr, err)
}
```

This is the SQL:
```sql
SELECT 
    accountbase.rowid, acctrounds.rnd, accountbase.data 
FROM acctrounds LEFT JOIN accountbase ON address=? 
WHERE id='acctbase'
```

If the account doesn't exist a record is returned anyway with `acctrounds.rnd`.
That causes `LookupAccount` to return with just the `Addr` set:
```go
data.Addr = addr
if len(buf) > 0 && rowid.Valid {
    data.Ref = sqlRowRef{rowid.Int64}
    err = protocol.Decode(buf, &data.AccountData)
    return err
}
// we don't have that account, just return the database round.
return nil
```

the exact same behavior the old code has with:
```go
a.updateOld(idx, trackerdb.PersistedAccountData{Addr: addr})
```

[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/acctdeltas.go:493]: not my thread but if account exist but it has nil data `en(buf) > 0 && rowid.Valid` prevents from setting `data.Ref`. In this case `data.oldAcct.Ref == nil` would indicate no old data but in fact there is an old data in DB. at this point two DBs with such invalid entry would diverge on updated and non-updated nodes.

File: ledger/acctdeltas.go (hunk 1 above)
Comment (algorandskiy) [reply ledger/acctdeltas.go:483]: a problem could be in a code that uses `Round=0` as some indication similarly to `rowid=0`. I'm pretty sure there is no such code, but have you checked? The PR description claimed the code is equivalent but it is not for this case and for the empty data as discussed in a thread below.

**Comment by algorandskiy** [comment]
I reexamined code paths and it appears even an empty/nil accounts in `accountbase` would not cause any problems because of:
1. LookupAccount would report not found => cannot update such account
2. Resources update would resolve address to rowid but such loaded account (really rowid only) does not appears in modification lists => no writes, no caching issues.

Having said that it appears safe to proceed with the change. As a safety measure there is a [PR](https://github.com/algorand/go-algorand/pull/5258) to clean up possible empty accounts but not sure if we even need this.

**Review by algorandskiy — APPROVED** [review]

**Comment by algorandskiy** [comment]
Here is a demo making an empty account and transferring to it. DB will fail to commit b/c of unique constraint error https://github.com/algorand/go-algorand/compare/master...algorandskiy:go-algorand:pavel/empty-acct-test-demo

File: ledger/store/trackerdb/sqlitedriver/sql.go [hunk 4]
```diff
@@ -476,6 +477,9 @@ func (qs *accountsDbQueries) LookupAccount(addr basics.Address) (data trackerdb.
 				data.Ref = sqlRowRef{rowid.Int64}
 				err = protocol.Decode(buf, &data.AccountData)
 				return err
+			} else if len(buf) == 0 && rowid.Valid {
+				// explicit condition: if the account row exists AND it has no data in it
+				logging.Base().Warnf("account %v exists but has no data in the database", addr)
```
Comment (algorandskiy) [inline ledger/store/trackerdb/sqlitedriver/sql.go:482]:
```suggestion
				logging.Base().Warnf("account %s exists but has no data in the database", addr)
```

**Review by algorandskiy — COMMENTED** [review]
Please change the address logging %v -> %s - we use canonical string addresses almost everywhere outside the code.
LGTM. The reason why I approve this is the resources migration converted all possible old empty/nil accounts, and no such records should be in the DB (that existed at the very early days).

**Review by algorandskiy — DISMISSED** [review]

**Comment by algorandskiy** [comment]
@cce please take another look.
There are two write paths for accountbase: accountsNewRound and catchpoints. The former one serialized data. The latter writes pre-serialized data but it decodes before, so no empty values possible.

File: ledger/store/trackerdb/data_test.go [hunk 5]
```diff
@@ -1252,3 +1252,17 @@ func TestBaseVotingDataReflect(t *testing.T) {
 
 	require.Equal(t, 7, reflect.TypeOf(BaseVotingData{}).NumField(), "update all getters and setters for baseVotingData and change the field count")
 }
+
+// TestBaseAccountDataDecodeEmpty ensures no surprises when decoding nil/empty data.
+func TestBaseAccountDataDecodeEmpty(t *testing.T) {
+	var b BaseAccountData
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/store/trackerdb/data_test.go:1267]: fixed

**Review by algorandskiy — DISMISSED** [review]

**Review by algorandskiy — APPROVED** [review]

File: ledger/acctdeltas.go (hunk 1 above)
Comment (bbroder-algo) [reply ledger/acctdeltas.go:483]: Hello, isn't the account data hashed by the catchpoint tracker, so wouldn't populating it with a new nonzero round number change that hash?

Comment (algorandskiy) [reply ledger/acctdeltas.go:483]: You are right Round might affect MT hash but we've checked all possible code paths and returning Round=0 not possible now.

File: ledger/acctdeltas.go [hunk 6]
```diff
@@ -476,29 +476,13 @@ func (a *compactAccountDeltas) accountsLoadOld(tx trackerdb.TransactionScope) (e
...
-			// Note: the err will be ignored in this case since `err` is being shadowed.
-			// this behaviour is equivalent to `err = nil`
-		default:
+		data, err := arw.LookupAccount(addr)
+		if err != nil {
 			// unexpected error - let the caller know that we couldn't complete the operation.
 			return err
 		}
+		// update the account
+		a.updateOld(idx, data)
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/acctdeltas.go:485]: Agreed, old Round is not used anywhere

**Comment by bbroder-algo** [comment]
what about here:
https://github.com/icorderi/go-algorand/blob/refactor/remove-LookupAccountDataByAddress/ledger/store/trackerdb/hashing.go#L131

**Comment by algorandskiy** [comment]
You mean MT deleting here, right?
```
		delta := accountsDeltas.getByIdx(i)
		if !delta.oldAcct.AccountData.IsEmpty() {
			deleteHash := trackerdb.AccountHashBuilderV6(delta.address, &delta.oldAcct.AccountData, protocol.Encode(&delta.oldAcct.AccountData))
			deleted, err := ct.balancesTrie.Delete(deleteHash)
```

**Comment by bbroder-algo** [comment]
confirmed that newAcct/oldAcct hashing is using BaseOnlineAccountData part of PersistedOnlineAccountData and not the round.
